File: java/src/basesort/BubbleSort.java
package basesort;

import java.util.Arrays;

public class BubbleSort {
    /*
     * 冒泡排序: 相互间两两比较，将大的移动到数组右端
     * 空间复杂度 O(1)、时间复杂度 O(n²)，
     * 是一种稳定的排序算法（稳定是指，两个相等的数，在排序过后，相对位置保持不变）
     */
    public static void bubbleSort(int[] nums) {

        int size = nums.length;
        // swapFlag 用于标识当此遍历是否进行交换，如果没有进行说明已排序完成，无需再进行下次遍历
        boolean swapFlag = true;
        for (int i = 0; (i < size - 1) && swapFlag; i++) {
            swapFlag = false;
            for (int j = 0; j < size - i - 1; j++) {
                if (nums[j] > nums[j + 1]) {
                    swap(nums, j, j + 1);
                    swapFlag = true;
                }
            }
        }
    }

    public static void swap(int[] nums, int index1, int index2) {
        int temp = nums[index1];
        nums[index1] = nums[index2];
        nums[index2] = temp;
    }

    public static void main(String[] args) {
        int[] testNums = {16,3,7,1,9,8,2,4};
        bubbleSort(testNums);
        System.out.println(Arrays.toString(testNums));
    }
}

File: cpp/base_sort/BubbleSort/BubbleSort.cpp
#include <iostream>
#include "BubbleSort.h"
using namespace std;

BubbleSort::BubbleSort(){}

void BubbleSort::bubbleSort(int nums[], int length) {
	bool swaped = true;
	for (int i = 0; i < length && swaped; i++) {
		swaped = false;
		for (int j = 1; j < length - i; j++) {
			if (nums[j-1] > nums[j]) {
				swap(nums[j-1], nums[j]);
				swaped = true;
			}
		}
	}
	for (int a = 0; a < length; a++){
		cout << nums[a] << endl;
	}
	
}

void BubbleSort::swap(int &a, int &b) {
	int temp = a;
	a = b;
	b = temp;
}

/*
int main() {
	BubbleSort sort;
	int input[7] = {10, 21, 9, 7, 1, 42, 37};
	int length = sizeof(input) / sizeof(input[0]);
	sort.bubbleSort(input, length);
}
*/